File: android/Reguerta/app/src/main/java/com/reguerta/user/data/freshness/FirestoreCriticalDataRefresher.kt
package com.reguerta.user.data.freshness

import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Source
import com.reguerta.user.data.access.decodeDirectoryMemberDocument
import com.reguerta.user.data.access.decodeMemberDocument
import com.reguerta.user.data.commitments.FirestoreSeasonalCommitmentRepository
import com.reguerta.user.data.firestore.ReguertaFirestoreCollection
import com.reguerta.user.data.firestore.ReguertaFirestoreEnvironment
import com.reguerta.user.data.firestore.ReguertaFirestorePath
import com.reguerta.user.data.firestore.toRepositoryException
import com.reguerta.user.data.products.decodeProductDocuments
import com.reguerta.user.domain.RepositoryErrorKind
import com.reguerta.user.domain.RepositoryException
import com.reguerta.user.domain.access.Member
import com.reguerta.user.domain.access.canManageMembers
import com.reguerta.user.domain.freshness.CriticalCollection
import com.reguerta.user.domain.freshness.CriticalDataRefreshPayload
import com.reguerta.user.domain.freshness.CriticalDataRefreshScope
import com.reguerta.user.domain.freshness.CriticalDataRefresher
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.tasks.await

class FirestoreCriticalDataRefresher(
    private val firestore: FirebaseFirestore,
) : CriticalDataRefresher {
    override suspend fun refresh(
        scope: CriticalDataRefreshScope,
        collections: Set<CriticalCollection>,
    ): CriticalDataRefreshPayload = coroutineScope {
        val environment = scope.resolveEnvironment()
        val path = ReguertaFirestorePath(environment = environment)
        val authenticatedMember = refreshMember(
            path = path,
            memberId = scope.authenticatedMemberId,
            resource = "criticalDataRefresh.members.authenticated",
        ).requiringPrincipal(scope.principalUid)
        if (scope.requiresAccessScopeRetry(authenticatedMember)) {
            return@coroutineScope CriticalDataRefreshPayload(
                authenticatedMemberId = authenticatedMember.id,
                authenticatedMember = authenticatedMember,
                selectedMember = null,
                seasonalCommitments = null,
                requiresAccessScopeRetry = true,
            )
        }

        val selectedMember = if (scope.memberId == authenticatedMember.id) {
            authenticatedMember
        } else {
            refreshMember(
                path = path,
                memberId = scope.memberId,
                resource = "criticalDataRefresh.members.selected",
            ).also { member ->
                listOf(member).requiringActiveSelectedMember(scope.memberId)
            }
        }
        val seasonalCommitments = async {
            FirestoreSeasonalCommitmentRepository(
                firestore = firestore,
                environment = environment,
            ).getActiveCommitmentsForMemberFromServer(selectedMember)
        }
        val pendingResults = collections.map { collection ->
            async {
                refreshCollection(
                    collection = collection,
                    scope = scope,
                    path = path,
                    environment = environment,
                    authenticatedMember = authenticatedMember,
                    selectedMember = selectedMember,
                )
            }
        }
        val results = pendingResults.awaitAll()

        CriticalDataRefreshPayload(
            authenticatedMemberId = authenticatedMember.id,
            authenticatedMember = authenticatedMember,
            selectedMember = selectedMember,
            seasonalCommitments = seasonalCommitments.await(),
            members = results.filterIsInstance<CriticalRefreshResult.Members>()
                .singleOrNull()
                ?.value,
            products = results.filterIsInstance<CriticalRefreshResult.Products>()
                .singleOrNull()
                ?.value,
        )
    }

    private suspend fun refreshCollection(
        collection: CriticalCollection,
        scope: CriticalDataRefreshScope,
        path: ReguertaFirestorePath,
        environment: ReguertaFirestoreEnvironment,
        authenticatedMember: Member,
        selectedMember: Member,
    ): CriticalRefreshResult = try {
        when (collection) {
            CriticalCollection.USERS -> CriticalRefreshResult.Members(
                refreshMembers(
                    scope = scope,
                    path = path,
                    authenticatedMember = authenticatedMember,
                    selectedMember = selectedMember,
                ),
            )

            CriticalCollection.PRODUCTS -> CriticalRefreshResult.Products(
                refreshProducts(path = path),
            )

            CriticalCollection.ORDERS -> {
                coroutineScope {
                    listOf(
                        path.collectionPath(ReguertaFirestoreCollection.ORDERS),
                        environment.legacyCollectionPath("orders"),
                    ).map { collectionPath ->
                        async {
                            refreshOwnerScopedCollection(
                                collectionPath = collectionPath,
                                memberId = scope.memberId,
                            )
                        }
                    }.awaitAll()
                }
                CriticalRefreshResult.Completed
            }

            CriticalCollection.ORDERLINES -> {
                coroutineScope {
                    listOf(
                        path.collectionPath(ReguertaFirestoreCollection.ORDER_LINES),
                        environment.legacyCollectionPath("orderLines"),
                    ).map { collectionPath ->
                        async {
                            refreshOwnerScopedCollection(
                                collectionPath = collectionPath,
                                memberId = scope.memberId,
                            )
                        }
                    }.awaitAll()
                }
                CriticalRefreshResult.Completed
            }

            CriticalCollection.CONTAINERS -> {
                firestore.collection(environment.legacyCollectionPath("containers"))
                    .get(Source.SERVER)
                    .await()
                CriticalRefreshResult.Completed
            }

            CriticalCollection.MEASURES -> {
                firestore.collection(environment.legacyCollectionPath("measures"))
                    .get(Source.SERVER)
                    .await()
                CriticalRefreshResult.Completed
            }
        }
    } catch (error: CancellationException) {
        throw error
    } catch (error: Exception) {
        throw error.toRepositoryException(resource = "criticalDataRefresh.${collection.wireKey}")
    }

    private suspend fun refreshMembers(
        scope: CriticalDataRefreshScope,
        path: ReguertaFirestorePath,
        authenticatedMember: Member,
        selectedMember: Member,
    ): List<Member> {
        val usersPath = path.collectionPath(ReguertaFirestoreCollection.USERS)
        if (scope.canManageMembers) {
            val snapshot = firestore.collection(usersPath).get(Source.SERVER).await()
            return snapshot.documents
                .map { document -> decodeMemberDocument(document.id, document.requiredData("members.document")) }
                .associateByTo(linkedMapOf(), Member::id)
                .apply {
                    this[authenticatedMember.id] = authenticatedMember
                    this[selectedMember.id] = selectedMember
                }
                .values
                .sortedBy { member -> member.displayName.lowercase() }
                .requiringActiveSelectedMember(scope.memberId)
        }

        val directoryPath = path.collectionPath(ReguertaFirestoreCollection.MEMBER_DIRECTORY)
        return firestore.collection(directoryPath)
            .whereEqualTo("isActive", true)
            .get(Source.SERVER)
            .await()
            .documents
            .map { document ->
                decodeDirectoryMemberDocument(
                    documentId = document.id,
                    data = document.requiredData("members.directory.document"),
                )
            }
            .associateByTo(linkedMapOf(), Member::id)
            .apply {
                this[authenticatedMember.id] = authenticatedMember
                this[selectedMember.id] = selectedMember
            }
            .values
            .sortedBy { member -> member.displayName.lowercase() }
            .requiringActiveSelectedMember(scope.memberId)
    }

    private suspend fun refreshMember(
        path: ReguertaFirestorePath,
        memberId: String,
        resource: String,
    ): Member = try {
        val usersPath = path.collectionPath(ReguertaFirestoreCollection.USERS)
        val document = firestore.document("$usersPath/$memberId")
            .get(Source.SERVER)
            .await()
        if (!document.exists()) {
            throw RepositoryException(
                kind = RepositoryErrorKind.NOT_FOUND,
                resource = resource,
            )
        }
        decodeMemberDocument(
            documentId = document.id,
            data = document.requiredData(resource),
        )
    } catch (error: CancellationException) {
        throw error
    } catch (error: Exception) {
        throw error.toRepositoryException(resource = resource)
    }

    private suspend fun refreshProducts(path: ReguertaFirestorePath) =
        firestore.collection(path.collectionPath(ReguertaFirestoreCollection.PRODUCTS))
            .get(Source.SERVER)
            .await()
            .documents
            .map { document -> document.id to document.requiredData("products.document") }
            .let(::decodeProductDocuments)

    private suspend fun refreshOwnerScopedCollection(
        collectionPath: String,
        memberId: String,
    ) = coroutineScope {
        OWNER_FIELD_NAMES.map { ownerField ->
            async {
                firestore.collection(collectionPath)
                    .whereEqualTo(ownerField, memberId)
                    .get(Source.SERVER)
                    .await()
            }
        }.awaitAll()
        Unit
    }
}

private sealed interface CriticalRefreshResult {
    data class Members(val value: List<Member>) : CriticalRefreshResult

    data class Products(val value: List<com.reguerta.user.domain.products.Product>) : CriticalRefreshResult

    data object Completed : CriticalRefreshResult
}

private val OWNER_FIELD_NAMES = listOf("userId", "memberId")

private fun CriticalDataRefreshScope.resolveEnvironment(): ReguertaFirestoreEnvironment =
    ReguertaFirestoreEnvironment.entries.firstOrNull { candidate ->
        candidate.wireValue == environment
    } ?: throw RepositoryException(
        kind = RepositoryErrorKind.INVALID_DATA,
        resource = "criticalDataRefresh.environment",
    )

private fun ReguertaFirestoreEnvironment.legacyCollectionPath(collection: String): String =
    "$wireValue/collections/$collection"

private fun DocumentSnapshot.requiredData(resource: String): Map<String, Any> =
    data ?: throw RepositoryException(
        kind = RepositoryErrorKind.INVALID_DATA,
        resource = resource,
    )

internal fun CriticalDataRefreshScope.requiresAccessScopeRetry(
    authenticatedMember: Member,
): Boolean = authenticatedMember.id != authenticatedMemberId ||
    !authenticatedMember.isActive ||
    authenticatedMember.canManageMembers != canManageMembers ||
    (!authenticatedMember.canManageMembers && memberId != authenticatedMemberId)

internal fun Member.requiringPrincipal(principalUid: String): Member {
    if (authUid != principalUid) {
        throw RepositoryException(
            kind = RepositoryErrorKind.PERMISSION_DENIED,
            resource = "criticalDataRefresh.members.authenticatedPrincipal",
        )
    }
    return this
}

internal fun List<Member>.requiringActiveSelectedMember(memberId: String): List<Member> {
    val selectedMember = firstOrNull { member -> member.id == memberId }
        ?: throw RepositoryException(
            kind = RepositoryErrorKind.NOT_FOUND,
            resource = "criticalDataRefresh.members.selected",
        )
    if (!selectedMember.isActive) {
        throw RepositoryException(
            kind = RepositoryErrorKind.PERMISSION_DENIED,
            resource = "criticalDataRefresh.members.selected",
        )
    }
    return this
}

File: ios/Reguerta/Reguerta/Data/Freshness/FirestoreCriticalDataRefresher.swift
import FirebaseFirestore
import Foundation

struct FirestoreCriticalDataRefresher: CriticalDataRefreshing {
    private let db: Firestore

    init(db: Firestore) {
        self.db = db
    }

    func refresh(
        collections: Set<CriticalCollection>,
        scope: CriticalDataRefreshScope
    ) async throws -> CriticalDataRefreshPayload {
        try Task.checkCancellation()
        let authenticatedMember = try await refreshMember(
            id: scope.authenticatedMemberID,
            scope: scope,
            resource: "criticalData.users.authenticatedMember"
        )
        try Task.checkCancellation()
        guard authenticatedMember.authUid == scope.principalUID else {
            throw RepositoryError.permissionDenied(
                resource: "criticalData.users.authenticatedMemberLink"
            )
        }
        guard authenticatedMember.canManageMembers == scope.canManageMembers else {
            return CriticalDataRefreshPayload(
                authenticatedMember: authenticatedMember,
                selectedMember: scope.authenticatedMemberID == scope.memberID
                    ? authenticatedMember
                    : nil
            )
        }
        let selectedMember = if scope.authenticatedMemberID == scope.memberID {
            authenticatedMember
        } else {
            try await refreshMember(
                id: scope.memberID,
                scope: scope,
                resource: "criticalData.users.selectedMember"
            )
        }
        try Task.checkCancellation()

        async let materializedState = refreshMaterializedState(
            collections: collections,
            scope: scope,
            authenticatedMember: authenticatedMember,
            selectedMember: selectedMember
        )
        async let cacheRefresh: Void = refreshCacheOnlyCollections(
            collections: collections,
            scope: scope
        )
        let (resolvedState, _) = try await (materializedState, cacheRefresh)
        try Task.checkCancellation()
        return CriticalDataRefreshPayload(
            authenticatedMember: authenticatedMember,
            selectedMember: selectedMember,
            members: resolvedState.members,
            products: resolvedState.products,
            seasonalCommitments: resolvedState.commitments
        )
    }

    private func refreshMaterializedState(
        collections: Set<CriticalCollection>,
        scope: CriticalDataRefreshScope,
        authenticatedMember: Member,
        selectedMember: Member
    ) async throws -> MaterializedCriticalDataState {
        async let members = collections.contains(.users)
            ? refreshMembers(
                scope: scope,
                authenticatedMember: authenticatedMember,
                selectedMember: selectedMember
            )
            : nil
        async let products = collections.contains(.products)
            ? refreshProducts(scope: scope)
            : nil
        async let commitments = refreshSeasonalCommitments(
            selectedMember: selectedMember,
            scope: scope
        )
        return try await MaterializedCriticalDataState(
            members: members,
            products: products,
            commitments: commitments
        )
    }

    private func refreshCacheOnlyCollections(
        collections: Set<CriticalCollection>,
        scope: CriticalDataRefreshScope
    ) async throws {
        async let orders: Void = collections.contains(.orders)
            ? refreshOwnedCollection(
                plusCollection: .orders,
                legacyCollectionName: "orders",
                scope: scope,
                resource: "criticalData.orders"
            )
            : ()
        async let orderlines: Void = collections.contains(.orderlines)
            ? refreshOwnedCollection(
                plusCollection: .orderlines,
                legacyCollectionName: "orderLines",
                scope: scope,
                resource: "criticalData.orderlines"
            )
            : ()
        async let containers: Void = collections.contains(.containers)
            ? refreshLegacyCollection(
                named: "containers",
                scope: scope,
                resource: "criticalData.containers"
            )
            : ()
        async let measures: Void = collections.contains(.measures)
            ? refreshLegacyCollection(
                named: "measures",
                scope: scope,
                resource: "criticalData.measures"
            )
            : ()
        _ = try await (orders, orderlines, containers, measures)
    }
}

private extension FirestoreCriticalDataRefresher {
    private func refreshMember(id: String, scope: CriticalDataRefreshScope, resource: String) async throws -> Member {
        let users = db.reguertaCollection(.users, environment: scope.environment)
        do {
            let snapshot = try await users.document(id)
                .getDocument(source: .server)
            guard snapshot.exists, let data = snapshot.data() else {
                throw RepositoryError.notFound(resource: resource)
            }
            let member = try FirestoreMemberRepository.member(
                documentID: snapshot.documentID,
                data: data
            )
            guard member.isActive else {
                throw RepositoryError.permissionDenied(resource: resource)
            }
            return member
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: resource)
        }
    }

    private func refreshMembers(
        scope: CriticalDataRefreshScope,
        authenticatedMember: Member,
        selectedMember: Member
    ) async throws -> [Member] {
        let visibleMembersQuery: Query = if scope.canManageMembers {
            db.reguertaCollection(.users, environment: scope.environment)
        } else {
            db.reguertaCollection(.memberDirectory, environment: scope.environment)
                .whereField("isActive", isEqualTo: true)
        }

        do {
            let visibleSnapshot = try await visibleMembersQuery.getDocuments(source: .server)
            let visibleMembers: [Member]

            if scope.canManageMembers {
                visibleMembers = try visibleSnapshot.documents.map { document in
                    try FirestoreMemberRepository.member(
                        documentID: document.documentID,
                        data: document.data()
                    )
                }
            } else {
                visibleMembers = try visibleSnapshot.documents.map { document in
                    try FirestoreMemberRepository.directoryMember(
                        documentID: document.documentID,
                        data: document.data()
                    )
                }
            }
            return (
                visibleMembers.filter {
                    $0.id != authenticatedMember.id && $0.id != selectedMember.id
                } + [authenticatedMember, selectedMember]
            )
                .reduce(into: [String: Member]()) { result, member in
                    result[member.id] = member
                }
                .values
                .sorted { lhs, rhs in
                    lhs.displayName.localizedCaseInsensitiveCompare(rhs.displayName) == .orderedAscending
                }
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "criticalData.users")
        }
    }

    private func refreshProducts(scope: CriticalDataRefreshScope) async throws -> [Product] {
        do {
            let snapshot = try await db
                .reguertaCollection(.products, environment: scope.environment)
                .getDocuments(source: .server)
            return try snapshot.documents.map { document in
                try FirestoreProductRepository.product(
                    documentID: document.documentID,
                    data: document.data()
                )
            }
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "criticalData.products")
        }
    }

    private func refreshSeasonalCommitments(
        selectedMember: Member,
        scope: CriticalDataRefreshScope
    ) async throws -> [SeasonalCommitment] {
        let repository = FirestoreSeasonalCommitmentRepository(
            db: db,
            environment: scope.environment
        )
        let commitmentsByLookup = try await withThrowingTaskGroup(
            of: [SeasonalCommitment].self
        ) { group in
            for lookupKey in selectedMember.seasonalCommitmentLookupKeys {
                group.addTask {
                    try await repository.activeCommitmentsFromServer(userId: lookupKey)
                }
            }
            var collected: [SeasonalCommitment] = []
            for try await commitments in group {
                collected.append(contentsOf: commitments)
            }
            return collected
        }
        return Dictionary(
            commitmentsByLookup.map { ($0.id, $0) },
            uniquingKeysWith: { _, latest in latest }
        ).values.sorted { lhs, rhs in
            if lhs.seasonKey.localizedCaseInsensitiveCompare(rhs.seasonKey) != .orderedSame {
                return lhs.seasonKey.localizedCaseInsensitiveCompare(rhs.seasonKey) == .orderedAscending
            }
            return lhs.productId.localizedCaseInsensitiveCompare(rhs.productId) == .orderedAscending
        }
    }

    private func refreshOwnedCollection(
        plusCollection: ReguertaFirestoreCollection,
        legacyCollectionName: String,
        scope: CriticalDataRefreshScope,
        resource: String
    ) async throws {
        let plusPath = ReguertaFirestorePath(environment: scope.environment)
            .collectionPath(plusCollection)
        let legacyPath = legacyCollectionPath(
            named: legacyCollectionName,
            environment: scope.environment
        )

        do {
            async let plusUserID: Void = refreshOwnedQuery(
                path: plusPath,
                ownerField: "userId",
                memberID: scope.memberID
            )
            async let plusMemberID: Void = refreshOwnedQuery(
                path: plusPath,
                ownerField: "memberId",
                memberID: scope.memberID
            )
            async let legacyUserID: Void = refreshOwnedQuery(
                path: legacyPath,
                ownerField: "userId",
                memberID: scope.memberID
            )
            async let legacyMemberID: Void = refreshOwnedQuery(
                path: legacyPath,
                ownerField: "memberId",
                memberID: scope.memberID
            )
            _ = try await (plusUserID, plusMemberID, legacyUserID, legacyMemberID)
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: resource)
        }
    }

    private func refreshOwnedQuery(path: String, ownerField: String, memberID: String) async throws {
        _ = try await db.collection(path)
            .whereField(ownerField, isEqualTo: memberID)
            .getDocuments(source: .server)
    }

    private func refreshLegacyCollection(
        named collectionName: String,
        scope: CriticalDataRefreshScope,
        resource: String
    ) async throws {
        do {
            _ = try await db.collection(
                legacyCollectionPath(named: collectionName, environment: scope.environment)
            ).getDocuments(source: .server)
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: resource)
        }
    }

    private func legacyCollectionPath(named collectionName: String, environment: SessionEnvironment) -> String {
        "\(environment.rawValue)/collections/\(collectionName)"
    }
}

private nonisolated struct MaterializedCriticalDataState: Sendable {
    let members: [Member]?
    let products: [Product]?
    let commitments: [SeasonalCommitment]
}
